pkg/services/librarypanels/database.go

```diff
@@ -104,3 +108,54 @@ func (lps *LibraryPanelService) getAllLibraryPanels(c *models.ReqContext) ([]Lib
 
 	return libraryPanels, err
 }
+
+// updateLibraryPanel updates a Library Panel.
+func (lps *LibraryPanelService) updateLibraryPanel(c *models.ReqContext, cmd updateLibraryPanelCommand) (LibraryPanel, error) {
+	var libraryPanel LibraryPanel
+	err := lps.SQLStore.WithTransactionalDbSession(context.Background(), func(session *sqlstore.DBSession) error {
+		panelInDb, err := getLibraryPanel(session, cmd.UID, c.SignedInUser.OrgId)
+		if err != nil {
+			return err
+		}
+
+		libraryPanel = LibraryPanel{
+			ID:        panelInDb.ID,
+			OrgID:     c.SignedInUser.OrgId,
+			FolderID:  cmd.FolderID,
+			UID:       cmd.UID,
+			Name:      cmd.Name,
+			Model:     cmd.Model,
+			Created:   panelInDb.Created,
+			CreatedBy: panelInDb.CreatedBy,
+			Updated:   time.Now(),
+			UpdatedBy: c.SignedInUser.UserId,
+		}
+
+		if cmd.FolderID == 0 {
+			libraryPanel.FolderID = panelInDb.FolderID
+		}
+		if cmd.Name == "" {
+			libraryPanel.Name = panelInDb.Name
+		}
+		if cmd.Model == nil {
+			libraryPanel.Model = panelInDb.Model
+		}
+
+		if res, err := session.Query("SELECT 1 FROM library_panel WHERE org_id=? AND folder_id=? AND name=? AND uid <>?",
+			libraryPanel.OrgID, libraryPanel.FolderID, libraryPanel.Name, libraryPanel.UID); err != nil {
+			return err
```

Comment:
Shouldn't this be handled via  UNIQUE constraint in the database? Checking this in the backend code looks really fragile to me. @sakjur @wbrowne thoughts?

Reply:
What you're saying sounds correct to me. Do the regular transaction semantics guarantee that you're not inserting "the same" row twice here without anything like a unique constraint?

```
   0
 /  \
A    B
 \  /
  AB
```

Reply:
I _think_ transaction semantics would also protect you here from inserting a non-unique row, but there might be edge cases/race conditions I'm not thinking of. I feel a bit rusty when it comes to considering transaction edge cases. The UNIQUE constraint should be a much safer bet in any case.

Reply:
This makes sense to me as well. Do we have a way of defining UNIQUE constraints using our current table definitions with xorm?

Reply:
It has to be added in your migrations I think.

Reply:
@aknuds1 is this what you mean by migrations? Because this doesn't seem to work :( but I'm probably doing something wrong.
```go
		Indices: []*migrator.Index{
			{Cols: []string{"org_id", "folder_id", "name"}, Type: migrator.UniqueIndex},
		}
```

Reply:
I'm fairly certain it should be done through migrations, but don't know the exact details of how we do it in Grafana since I haven't added unique constraints there. There should be several equivalent ways of defining uniqueness in SQL as I recall. Can you find examples of it in our migrations?

Reply:
Reading up on this here https://github.com/go-xorm/manual-en-US/blob/master/chapter-02/4.columns.md

Reply:
It's probably wise to look for examples in our migrations also.

Reply:
@hugohaggmark What I've run into before: the index needs it's own IndexMigrator, it won't be picked up by the table migrator.

something like

```go
	mg.AddMigration(indexIdentifier string, migrator.NewAddIndexMigration(migrator.Table, *migrator.Index))
```

Reply:
Yupp, I'm just about to try that. Thank you @sakjur

Reply:
Thanks @sakjur for chiming in :+1:

Reply:
That worked beautifully 🎉

Reply:
Added `UniqueIndex` and changed both `patchLibraryPanel` and `createLibraryPanel` functions to support this instead.